### PR TITLE
Allow Camel modules to gracefully handle missing dependencies

### PR DESCRIPTION
**Pull request by ctron**
This change does allow different camel functions to handle missing dependencies gracefully by waiting for them.

### CODE REVIEW

File: kura/org.eclipse.kura.camel.cloud.factory/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -12,6 +12,7 @@ Import-Package: org.apache.camel;version="[2.17.0,3.0.0)",
  org.apache.camel.model;version="[2.17.0,3.0.0)",
  org.apache.camel.support;version="[2.17.0,3.0.0)",
  org.eclipse.kura;version="[1.2,2.0)",
+ org.eclipse.kura.camel.bean;version="[1.1.0,2.0.0)",
```
Comment (MMaiero) [inline kura/org.eclipse.kura.camel.cloud.factory/META-INF/MANIFEST.MF]: Version interval with micro. It is different compared to the following imports. We discussed about that also for cloud stuff.

**Comment by MMaiero** [comment]
Trying to start Kura I end up with the following exception. Is it the supposed behavior?

```
2016-11-18 09:20:29,988 [Component Resolve Thread (Bundle 7)] WARN  o.e.k.c.r.CamelRunner - Failed to start context
org.apache.camel.FailedToCreateRouteException: Failed to create route route1: Route(route1)[[From[kura-cloud:myapp/xmltopic]] -> [To[log:M... because of Failed to resolve endpoint: kura-cloud://myapp/xmltopic due to: No component found with scheme: kura-cloud
	at org.apache.camel.model.RouteDefinition.addRoutes(RouteDefinition.java:201)
	at org.apache.camel.impl.DefaultCamelContext.startRoute(DefaultCamelContext.java:974)
	at org.apache.camel.impl.DefaultCamelContext.startRouteDefinitions(DefaultCamelContext.java:3301)
	at org.apache.camel.impl.DefaultCamelContext.doStartCamel(DefaultCamelContext.java:3024)
	at org.apache.camel.impl.DefaultCamelContext.access$000(DefaultCamelContext.java:175)
	at org.apache.camel.impl.DefaultCamelContext$2.call(DefaultCamelContext.java:2854)
	at org.apache.camel.impl.DefaultCamelContext$2.call(DefaultCamelContext.java:2850)
	at org.apache.camel.impl.DefaultCamelContext.doWithDefinedClassLoader(DefaultCamelContext.java:2873)
	at org.apache.camel.impl.DefaultCamelContext.doStart(DefaultCamelContext.java:2850)
	at org.apache.camel.support.ServiceSupport.start(ServiceSupport.java:61)
	at org.apache.camel.impl.DefaultCamelContext.start(DefaultCamelContext.java:2819)
	at org.eclipse.kura.camel.runner.CamelRunner.startCamel(CamelRunner.java:491)
	at org.eclipse.kura.camel.runner.CamelRunner$4.ready(CamelRunner.java:442)
	at org.eclipse.kura.camel.runner.DependencyRunner.triggerUpdate(DependencyRunner.java:113)
	at org.eclipse.kura.camel.runner.DependencyRunner.update(DependencyRunner.java:100)
	at org.eclipse.kura.camel.runner.DependencyRunner.access$0(DependencyRunner.java:91)
	at org.eclipse.kura.camel.runner.DependencyRunner$1.run(DependencyRunner.java:45)
	at org.eclipse.kura.camel.runner.DefaultServiceDependency$HandleImpl.triggerUpdate(DefaultServiceDependency.java:78)
	at org.eclipse.kura.camel.runner.DefaultServiceDependency$HandleImpl.adding(DefaultServiceDependency.java:65)
	at org.eclipse.kura.camel.runner.DefaultServiceDependency$HandleImpl$1.addingService(DefaultServiceDependency.java:39)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:932)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:894)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:107)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.dispatchEvent(BundleContextImpl.java:861)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:819)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:771)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:214)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.registerService(BundleContextImpl.java:433)
	at org.eclipse.equinox.internal.ds.InstanceProcess.registerService(InstanceProcess.java:536)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:260)
	at org.eclipse.equinox.internal.ds.Resolver.getEligible(Resolver.java:343)
	at org.eclipse.equinox.internal.ds.SCRManager.serviceChanged(SCRManager.java:222)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:107)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.dispatchEvent(BundleContextImpl.java:861)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:819)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:771)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:214)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.registerService(BundleContextImpl.java:433)
	at org.eclipse.equinox.internal.ds.InstanceProcess.registerService(InstanceProcess.java:536)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:260)
	at org.eclipse.equinox.internal.ds.Resolver.getEligible(Resolver.java:343)
	at org.eclipse.equinox.internal.ds.SCRManager.serviceChanged(SCRManager.java:222)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:107)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.dispatchEvent(BundleContextImpl.java:861)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:819)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:771)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:214)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.registerService(BundleContextImpl.java:433)
	at org.eclipse.equinox.internal.ds.InstanceProcess.registerService(InstanceProcess.java:536)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:260)
	at org.eclipse.equinox.internal.ds.Resolver.getEligible(Resolver.java:343)
	at org.eclipse.equinox.internal.ds.SCRManager.serviceChanged(SCRManager.java:222)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:107)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.dispatchEvent(BundleContextImpl.java:861)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:819)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:771)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:214)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.registerService(BundleContextImpl.java:433)
	at org.eclipse.equinox.internal.ds.InstanceProcess.registerService(InstanceProcess.java:536)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:260)
	at org.eclipse.equinox.internal.ds.Resolver.getEligible(Resolver.java:343)
	at org.eclipse.equinox.internal.ds.SCRManager.serviceChanged(SCRManager.java:222)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:107)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.dispatchEvent(BundleContextImpl.java:861)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:819)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:771)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:214)
	at org.eclipse.osgi.framework.internal.core.BundleContextImpl.registerService(BundleContextImpl.java:433)
	at org.eclipse.equinox.internal.ds.InstanceProcess.registerService(InstanceProcess.java:536)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:260)
	at org.eclipse.equinox.internal.ds.Resolver.buildNewlySatisfied(Resolver.java:473)
	at org.eclipse.equinox.internal.ds.Resolver.enableComponents(Resolver.java:217)
	at org.eclipse.equinox.internal.ds.SCRManager.performWork(SCRManager.java:816)
	at org.eclipse.equinox.internal.ds.SCRManager$QueuedJob.dispatch(SCRManager.java:783)
	at org.eclipse.equinox.internal.ds.WorkThread.run(WorkThread.java:89)
	at org.eclipse.equinox.internal.util.impl.tpt.threadpool.Executor.run(Executor.java:70)
Caused by: org.apache.camel.ResolveEndpointFailedException: Failed to resolve endpoint: kura-cloud://myapp/xmltopic due to: No component found with scheme: kura-cloud
	at org.apache.camel.impl.DefaultCamelContext.getEndpoint(DefaultCamelContext.java:594)
	at org.apache.camel.util.CamelContextHelper.getMandatoryEndpoint(CamelContextHelper.java:79)
	at org.apache.camel.model.RouteDefinition.resolveEndpoint(RouteDefinition.java:211)
	at org.apache.camel.impl.DefaultRouteContext.resolveEndpoint(DefaultRouteContext.java:107)
	at org.apache.camel.impl.DefaultRouteContext.resolveEndpoint(DefaultRouteContext.java:113)
	at org.apache.camel.model.FromDefinition.resolveEndpoint(FromDefinition.java:69)
	at org.apache.camel.impl.DefaultRouteContext.getEndpoint(DefaultRouteContext.java:89)
	at org.apache.camel.model.RouteDefinition.addRoutes(RouteDefinition.java:1052)
	at org.apache.camel.model.RouteDefinition.addRoutes(RouteDefinition.java:196)
	... 93 more
```

File: kura/org.eclipse.kura.camel.cloud.factory/META-INF/MANIFEST.MF (hunk 1 above)
Comment (ctron) [inline kura/org.eclipse.kura.camel.cloud.factory/META-INF/MANIFEST.MF]: Sure, did you create that ruleset already?

**Comment by ctron** [comment]
No. But it would be great to know what you did.

**Comment by MMaiero** [comment]
just built, installed in a raspberry and started

**Comment by MMaiero** [comment]
Did you manage here? Do you want it to be merged anyway and solve later?

**Comment by ctron** [comment]
Looks good for me. Let's merge it and fix it up something comes up.

**Comment by ctron** [comment]
Have you deleted the default Kura cloud client?

**Comment by MMaiero** [comment]
?

**Comment by ctron** [comment]
Have you, or have you not? That could explain it.

**Comment by MMaiero** [comment]
No, I've not.

**Comment by dwoodard1** [comment]
@MMaiero I just tried a fresh install on a RPi 2 and did not see the error. I will try to investigate Camel later today, but Kura started without exceptions for me.

**Comment by MMaiero** [comment]
@dwoodard1 I've tried several times and seems to happen few times, not all. I've defined a docker container (I don't have a rpi with me now) and out of 4 starts it happened 2 times. I've also built the distribution from scratch.
Could you try this [1] and see? I want to figure out if it's something in my environment that does not work. Thanks.

[1] https://s3-us-west-2.amazonaws.com/kura-folder/kura_2.1.0_raspberry-pi-2-3-nn_installer.deb

**Comment by dwoodard1** [comment]
Yes, I will try later today. Did you completely remove ~/.m2/repository/org/eclipse/kura before building the release branch? I think there are some potential conflicts if you have also built the develop branch.

When you say you have tried several times, are you re-installing each time or just restarting?


-------- Original message --------

**Comment by ctron** [comment]
So this seems like some sort of race condition. So one scenario which could be is that the Kura cloud service is missing (due to a late startup or something different) and then the "kura-cloud" component would be missing.

I made a small PR #802 which could fix this, if this is really the case. However I can only test this on Monday.

Thanks for looking into this.

**Comment by MMaiero** [comment]
> Did you completely remove ~/.m2/repository/org/eclipse/kura before building the release branch?

@dwoodard1 Yes, I'm always doing that since, in this period, I'm continuously switching between projects and branches.

> When you say you have tried several times, are you re-installing each time or just restarting?

Just to be sure, I've run them with a new docker image instance every time.

@ctron Thanks for the PR. I'll be able to look at it only on Monday.

**Pull request merged**